### PR TITLE
[BugFix][Performance] Avoid dtype-converting valid token count copies

### PR DESCRIPTION
**Pull request by Republix-ch**
### What this PR does / why we need it?

This is a follow-up to #10205.

#10205 changed `self.valid_sampled_token_count_cpu` to `torch.int64` to avoid an implicit dtype conversion during the asynchronous D2H copy for the MTP path. However, the shared copy helper also receives `torch.int32` ‎`valid_sampled_tokens_count` from other proposers:

- [triton_ngram_spec_decode](https://github.com/vllm-project/vllm-ascend/blob/9c3cf949db9dd352d21c9d4d31468d2958e0a953/vllm_ascend/ops/triton/spec_decode/ngram.py#L228-L259) returns `valid_sampled_tokens_count` as a view of the `torch.int32` output workspace:
    ```python
  base = torch.empty(..., dtype=torch.int32, device=device)
  valid_sampled_tokens_count = base[2 * batch_size : 3 * batch_size]
  ```
- [AscendExtractHiddenStatesProposer.prepare_next_token_ids_padded](https://github.com/vllm-project/vllm-ascend/blob/9c3cf949db9dd352d21c9d4d31468d2958e0a953/vllm_ascend/spec_decode/extract_hidden_states_proposer.py#L193-L198) explicitly converts the count to `torch.int32`:
   ```python
   valid_sampled_tokens_count = is_valid.to(torch.int32)
   ```
- These two return values are passed directly to `_copy_valid_sampled_token_count` at the [N-gram call site](https://github.com/vllm-project/vllm-ascend/blob/9c3cf949db9dd352d21c9d4d31468d2958e0a953/vllm_ascend/worker/model_runner_v1.py#L1744-L1760) and the [ExtractHiddenStates call site](https://github.com/vllm-project/vllm-ascend/blob/9c3cf949db9dd352d21c9d4d31468d2958e0a953/vllm_ascend/worker/model_runner_v1.py#L1805-L1814), respectively.

Those paths can therefore still trigger a dtype-converting D2H copy, which forces a host synchronization and defeats the side-stream overlap.

This PR normalizes the count tensor to the host buffer dtype on the NPU copy stream before calling `copy_(..., non_blocking=True)`. The original device tensor is retained for GPU-side async-spec-decode correction.

A regression test covers both `int32 -> int64` and `int64 -> int32` source/target dtype combinations and verifies that the tensor passed to the D2H copy matches the host buffer dtype.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- Added `TestCopyValidSampledTokenCount::test_normalizes_device_counts_to_host_buffer_dtype` covering both `int32 -> int64` and `int64 -> int32` source/target combinations.
- Relevant speculative-decoding E2E coverage is selected by CI.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

File: vllm_ascend/worker/model_runner_v1.py [hunk 1]
```diff
@@ -1434,11 +1434,13 @@ def _copy_valid_sampled_token_count(
...
         default_stream = torch.npu.current_stream()
-        with torch.npu.stream(self.valid_sampled_token_count_copy_stream): 
+        with torch.npu.stream(self.valid_sampled_token_count_copy_stream):
             self.valid_sampled_token_count_copy_stream.wait_stream(default_stream)
-            counts = valid_sampled_tokens_count
             counts_cpu = self.valid_sampled_token_count_cpu
             assert counts_cpu is not None
+            # Proposers may return different integer dtypes. Normalize on device
+            # because a dtype-converting D2H copy_ forces a host sync.
+            counts = valid_sampled_tokens_count.to(dtype=counts_cpu.dtype)
```
[the remark this answers is not in the document]
Comment (Republix-ch) [reply vllm_ascend/worker/model_runner_v1.py:1443]: The current head commit `eabd8fd9f` already contains the required trailer:

Signed-off-by: alex_chenghao <alex_chenghao@163.com>

The DCO check is also passed.

File: tests/ut/worker/a2/test_model_runner_v1.py [hunk 2]
```diff
@@ -882,6 +882,45 @@ def test_execute_model_ignores_need_timing_when_profiling_chunk_is_disabled(
...
+            with self.subTest(device_dtype=device_dtype, host_dtype=host_dtype):
+                runner = NPUModelRunner.__new__(NPUModelRunner)
+                runner.valid_sampled_token_count_event = MagicMock()
+                runner.valid_sampled_token_count_copy_stream = MagicMock()
+                host_buffer = self._HostBufferSpy(torch.empty(2, dtype=host_dtype))
+                runner.valid_sampled_token_count_cpu = host_buffer
+                runner.use_async_spec_decode = False
+                runner.input_batch = SimpleNamespace(prev_sampled_token_ids=None)
+
+                device_counts = torch.tensor([1, 3], dtype=device_dtype)
```
[the remark this answers is not in the document]
Comment (Republix-ch) [reply tests/ut/worker/a2/test_model_runner_v1.py:914]: This PR extends #10205 to the remaining int32 producer paths while the host buffer is int64. The unit test ensures that `copy_` receives matching dtypes. Since the synchronization mechanism is unchanged, an additional NPU performance test should not be necessary.

**Comment by Republix-ch** [comment]
Hello @MengqingCao @nwpu-zxr @linfeng-yuan, this PR is a follow-up to https://github.com/vllm-project/vllm-ascend/pull/10205. It handles the remaining dtype-mismatched D2H copies when speculative decoding paths return int32 token counts.

Could you please take a look? The selected E2E tests are also waiting for the `ready` label. Thanks!

**Comment by Republix-ch** [comment]
Hi @Ronald1995 @weijinqian0, gentle ping on this PR. As a small follow-up to #10205, this PR casts the count tensor to the host buffer’s dtype on device before issuing the asynchronous D2H copy. Could you please take a look when you have time? Thanks!
